### ISSUE TITLE
Refactor segment_identity

### ISSUE THREAD

**Issue by pwoods25443** [GlobalFishingWatch/pipe-segment#65]
Several things that need to be done to improve segment_identity

- use a sliding 30-day window instead of calendar month to assign identity to segments
- preserve most common non-normalized name and callsign
- include vessel type
- include "known imo" filter on imo number

### Design notes

- Create a daily table of counts of identity fields per seg_id `segment_identity_daily`
- Aggregate a 30-day window from `segment_identity_daily` into another daily table `segment_identity_window`.   Generate vessel_id in this table according to the current algorithm
- Aggregate across the entire time range from `segment_identity_window` to create 3 info tables
  * `segment_info` one row per seg_id
  * `vessel_info` one row per vessel_id
  * `segment_vessel` many yo many mapping of seg_id to vessel_id

**Comment by pwoods25443** [comment]
### WIP

```sql
CREATE TEMP FUNCTION mostCommon(arr ARRAY<STRUCT<value STRING, count INT64>>) AS (
  (
   select as struct
     value, 
     count,
     count/(select sum(r.count) from unnest(arr) as r ) as freq
   from (
     select as struct r.value as value, sum(r.count) as count from unnest(arr) as r group by r.value order by count DESC limit 1 
     )
   )
);

WITH
  messages AS (
  SELECT
    ssvid,
    seg_id,
    shipname,
    callsign,
    imo,
    lat
  FROM
    `world-fishing-827.pipe_production_b.messages_segmented_20180821`),
  segment_summary AS (
  SELECT
    ssvid,
    seg_id,
    COUNT(*) AS msg_count,
    COUNTIF(lat IS NOT NULL) AS pos_count
  FROM
    messages
  GROUP BY
    ssvid,
    seg_id ),

  shipname AS (
  SELECT
    seg_id,
    shipname as value,
    COUNT(*) AS count
  FROM
    messages
  WHERE
    shipname IS NOT NULL
  GROUP BY
    seg_id,
    shipname ),

  test_segs as (
  select
    * 
  from unnest(
    [
      STRUCT(
        'seg1' as seg_id, 
        [
          STRUCT('a' as value, 1 as count),
          STRUCT('b' as value, 2 as count)
        ] as shipname,
        [
          STRUCT('A123' as value, 1 as count),
          STRUCT('B123' as value, 2 as count)
        ] as callsign
      ),
      STRUCT(
        'seg1' as seg_id, 
        [
          STRUCT('a' as value, 2 as count),
          STRUCT('c' as value, 4 as count)
        ] as shipname,
        [
          STRUCT('A123' as value, 2 as count),
          STRUCT('C123' as value, 4 as count)
        ] as callsign
      ),
      STRUCT(
        'seg2' as seg_id, 
        [
          STRUCT('a' as value, 7 as count),
          STRUCT('c' as value, 1 as count)
        ] as shipname,
        [
          STRUCT('A123' as value, 7 as count),
          STRUCT('C123' as value, 1 as count)
        ] as callsign
      )
      
    ]
    )
  ),
  
  test_segs_old AS (
  SELECT
    'seg1' AS seg_id,
    'ssvid' AS ssvid,
    ARRAY(
    SELECT
      AS STRUCT 'a' AS value,
      1 AS count
    UNION ALL
    SELECT
      AS STRUCT 'b' AS value,
      2 AS count ) AS shipname,
    ARRAY(
    SELECT
      AS STRUCT 'A123' AS value,
      1 AS count
    UNION ALL
    SELECT
      AS STRUCT 'B123' AS value,
      2 AS count ) AS callsign
  UNION ALL
  SELECT
    'seg1' AS seg_id,
    'ssvid' AS ssvid,
    ARRAY(
    SELECT
      AS STRUCT 'a' AS value,
      2 AS count
    UNION ALL
    SELECT
      AS STRUCT 'c' AS value,
      4 AS count ) AS shipname,
    ARRAY(
    SELECT
      AS STRUCT 'A123' AS value,
      2 AS count
    UNION ALL
    SELECT
      AS STRUCT 'C123' AS value,
      4 AS count ) AS callsign
  UNION ALL
  SELECT
    'seg2' AS seg_id,
    'ssvid' AS ssvid,
    ARRAY(
    SELECT
      AS STRUCT 'a' AS value,
      7 AS count
    UNION ALL
    SELECT
      AS STRUCT 'c' AS value,
      1 AS count ) AS shipname,
    ARRAY(
    SELECT
      AS STRUCT 'A123' AS value,
      7 AS count
    UNION ALL
    SELECT
      AS STRUCT 'C123' AS value,
      1 AS count ) AS callsign ),
  --   test_consolidate as (
    --     select
    --       seg_id,
    --       ARRAY_CONCAT_AGG(shipname) as shipname,
    --       ARRAY_CONCAT_AGG(callsign) as callsign
    --     from test_segs
    --     group by seg_id
    --   ),
  --   seg_shipname as (
    --     select seg_id, shipname
    --     from test_segs
    --     cross join unnest(test_segs.shipname) as shipname
    --   ),
  --   test_agg as (
    --     select seg_id, shipname, sum(count) as count
    --     from test_unnest
    --     group by seg_id, shipname
    --   ),
--   shipname_count AS (
--   SELECT
--     seg_id,
--     STRUCT( shipname.value AS value,
--       SUM(shipname.count) OVER (PARTITION BY seg_id, shipname.value) AS count,
--       SUM(shipname.count) OVER (PARTITION BY seg_id) AS total ) AS shipname
--   FROM
--     test_segs
--   CROSS JOIN
--     UNNEST(test_segs.shipname) AS shipname ),
  --   shipname_top as (
    --     select
    --       seg_id,
    --       FIRST_VALUE(shipname) OVER(PARTITION BY seg_id ORDER BY shipname.count DESC) as shipname
    --     FROM shipname_count
    --   ),
 
  test_messages as (
    select 
      *
    from
      unnest ([
        STRUCT(1 as day, 'seg1' as seg_id, 'a' as shipname, 'a123' as callsign),
        STRUCT(1 as day, 'seg1' as seg_id, 'a' as shipname, 'a123' as callsign),
        STRUCT(1 as day, 'seg1' as seg_id, 'b' as shipname, 'b123' as callsign),
        STRUCT(1 as day, 'seg2' as seg_id, 'a' as shipname, 'a123' as callsign),
        STRUCT(1 as day, 'seg2' as seg_id, 'c' as shipname, 'c123' as callsign),
        STRUCT(1 as day, 'seg2' as seg_id, 'c' as shipname, 'c123' as callsign),
        STRUCT(1 as day, 'seg2' as seg_id, null as shipname, 'c123' as callsign),
        STRUCT(1 as day, 'seg3' as seg_id, 'd' as shipname, null as callsign),
        STRUCT(1 as day, 'seg3' as seg_id, 'd' as shipname, null as callsign),
        STRUCT(1 as day, 'seg3' as seg_id, 'd' as shipname, 'D123' as callsign),
        STRUCT(2 as day, 'seg1' as seg_id, 'a' as shipname, 'a123' as callsign),
        STRUCT(2 as day, 'seg1' as seg_id, 'a' as shipname, 'a123' as callsign),
        STRUCT(2 as day, 'seg1' as seg_id, 'b' as shipname, 'b123' as callsign),
        STRUCT(2 as day, 'seg2' as seg_id, 'a' as shipname, 'a123' as callsign),
        STRUCT(2 as day, 'seg2' as seg_id, 'c' as shipname, 'c123' as callsign),
        STRUCT(2 as day, 'seg2' as seg_id, 'c' as shipname, 'c123' as callsign),
        STRUCT(2 as day, 'seg4' as seg_id, null as shipname, null as callsign)
       
      ]) as message
  ),
  seg_sumary as (
    select 
      day, 
      seg_id,
      count(*) as msg_count
    from 
      test_messages
    group by 
      day, seg_id
  ),
  seg_shipname as (
    select 
      day, 
      seg_id, 
      array_agg(struct(value, count)) as shipname
    from (
        select
        day, seg_id,
        shipname as value, 
        count(shipname) as count
      from
        test_messages
      where shipname is not null
      group by day, seg_id, shipname
    )
    group by day, seg_id
  ),
  seg_callsign as (
    select 
      day, seg_id, 
      array_agg(struct(value, count)) as callsign
    from (
        select
        day, seg_id,
        callsign as value, 
        count(callsign) as count
      from
        test_messages
      where 
        callsign is not null
      group by day, seg_id, callsign
    )
    group by day, seg_id
  ),  
  daily_seg as (
    select
      *
    from
      seg_sumary
      left join seg_shipname using (day, seg_id)
      left join seg_callsign using (day, seg_id)
  ),
  seg_multi_day as (
    select 
      seg_id, 
      array_concat_agg(shipname) as shipname,
      array_concat_agg(callsign) as callsign
    from daily_seg
    group by seg_id
  ),
  seg_most_common as (
    select 
      seg_id, 
      mostCommon(shipname) as shipname,
      mostCommon(callsign) as callsign
    from
      seg_multi_day
  )



SELECT
  *
FROM
  seg_most_common
```

**Comment by pwoods25443** [comment]
Closed by #66
